--- a/runtime/js/mlBytes.js
+++ b/runtime/js/mlBytes.js
@@ -504,6 +504,7 @@
 }

 //Provides: caml_create_bytes const
+//Alias: caml_create_local_bytes
 //Requires: MlBytes,caml_invalid_argument
 function caml_create_bytes(len) {
   if (len < 0) caml_invalid_argument("Bytes.create");
--- a/runtime/wasm/string.wat
+++ b/runtime/wasm/string.wat
@@ -121,7 +121,8 @@

    (@string $Bytes_create "Bytes.create")

-   (func (export "caml_create_bytes")
+   (export "caml_create_local_bytes" (func $caml_create_bytes))
+   (func $caml_create_bytes (export "caml_create_bytes")
       (param $len (ref eq)) (result (ref eq))
       (local $l i32)
       (local.set $l (i31.get_s (ref.cast (ref i31) (local.get $len))))
