diff --git a/otherlibs/stdune/src/queue.ml b/otherlibs/stdune/src/queue.ml
index 6a5fbb493..fb3db35a9 100644
--- a/otherlibs/stdune/src/queue.ml
+++ b/otherlibs/stdune/src/queue.ml
@@ -1,10 +1,16 @@
 include Stdlib.Queue
 
+let create () = create ()
 let push t x = add x t
 let peek_exn t = peek t
 let pop_exn t = pop t
 let pop t = if is_empty t then None else Some (pop_exn t)
 let peek t = if is_empty t then None else Some (peek t)
+let clear t = clear t
+let copy t = copy t
+let is_empty t = is_empty t
+let length t = length t
 let iter t ~f = iter f t
 let fold t ~f ~init = fold f init t
+let transfer t1 t2 = transfer t1 t2
 let to_list t = List.rev (fold t ~f:(fun acc a -> a :: acc) ~init:[])
diff --git a/src/csexp_rpc/csexp_rpc.ml b/src/csexp_rpc/csexp_rpc.ml
index 0a5d308c1..1c1485d9c 100644
--- a/src/csexp_rpc/csexp_rpc.ml
+++ b/src/csexp_rpc/csexp_rpc.ml
@@ -236,10 +236,10 @@ module Session = struct
 
   external send : Unix.file_descr -> Bytes.t -> int -> int -> int = "dune_send"
 
-  let write =
+  let write t b =
     match Platform.OS.value with
-    | Linux -> send
-    | _ -> Unix.single_write
+    | Linux -> send t b
+    | _ -> Unix.single_write t b
   ;;
 
   let rec csexp_write_loop fd out_buf token =
