--- a/lib/automata.mli
+++ b/lib/automata.mli
@@ -1,3 +1,5 @@
+@@ portable
+
 (*
    RE - A regular expression library
 
@@ -20,10 +22,12 @@
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *)
 
+open Import
+
 (* Regular expressions *)
 
 module Mark : sig
-  type t [@@immediate]
+  type t : immediate
 
   val compare : t -> t -> int
   val start : t
@@ -54,7 +58,7 @@ module Rep_kind : sig
   val pp : t Fmt.t
 end
 
-type expr
+type expr : immutable_data
 
 val is_eps : expr -> bool
 val pp : expr Fmt.t
@@ -83,7 +87,7 @@ val rename : Ids.t -> expr -> expr
 (* States of the automata *)
 
 module Idx : sig
-  type t
+  type t : immediate
 
   val to_int : t -> int
 end
@@ -96,7 +100,7 @@ module Status : sig
 end
 
 module State : sig
-  type t
+  type t : immutable_data
 
   val pp : t Fmt.t
   val dummy : t
@@ -114,7 +118,7 @@ end
 (* Computation of the states following a given state *)
 
 module Working_area : sig
-  type t
+  type t : mutable_data
 
   val create : unit -> t
   val index_count : t -> int
