--- a/lib/color_map.mli
+++ b/lib/color_map.mli
@@ -1,3 +1,5 @@
+@@ portable
+
 (* Color maps exists to provide an optimization for the regex engine. The fact
    that some characters are entirely equivalent for some regexes means that we
    can use them interchangeably.
@@ -5,17 +7,17 @@
    A color map assigns a color to every character in our character set. Any two
    characters with the same color will be treated equivalently by the automaton.
 *)
-type t
+type t : mutable_data
 
 module Repr : sig
-  type t
+  type t : immutable_data
 
   val repr : t -> Cset.c -> char
   val length : t -> int
 end
 
 module Table : sig
-  type t
+  type t : immutable_data
 
   val get_char : t -> Cset.c -> char
   val get : t -> char -> Cset.c
