--- a/compiler/lib/ocaml_compiler.ml
+++ b/compiler/lib/ocaml_compiler.ml
@@ -21,8 +21,14 @@
 let rec constant_of_const c : Code.constant =
   let open Lambda in
   match c with
-  | Const_base (Const_int i) -> Int (Targetint.of_int_warning_on_overflow i)
-  | Const_base (Const_char c) -> Int (Targetint.of_int_exn (Char.code c))
+  | Const_base
+      (Const_int8 i | Const_untagged_int8 i |
+       Const_int16 i | Const_untagged_int16 i |
+       Const_int i | Const_untagged_int i) ->
+    Int (Targetint.of_int_warning_on_overflow i)
+
+  | Const_base (Const_char c | Const_untagged_char c) ->
+    Int (Targetint.of_int_exn (Char.code c))
   | Const_base (Const_string (s, _, _)) -> String s
   | Const_base (Const_float s | Const_unboxed_float s) -> Float (Int64.bits_of_float (float_of_string s))
   | Const_base (Const_float32 s | Const_unboxed_float32 s) -> Float32 (Int64.bits_of_float (Float32.of_string s |> Float32.to_float))
