--- a/src/spawn_stubs.c
+++ b/src/spawn_stubs.c
@@ -1,5 +1,9 @@
 #define _GNU_SOURCE
 
+#define CAML_INTERNALS
+/* for [caml_convert_signal_number] */
+#include <caml/signals.h>
+#undef CAML_INTERNALS
 #include <caml/mlvalues.h>
 #include <caml/memory.h>
 #include <caml/alloc.h>
@@ -8,11 +12,6 @@
 
 #include <errno.h>
 
-#define CAML_INTERNALS
-/* for [caml_convert_signal_number] */
-#include <caml/signals.h>
-#undef CAML_INTERNALS
-
 #if defined(__APPLE__)
 
 CAMLprim value spawn_is_osx()
